src/exo/LoopIR_scheduling.py

```diff
@@ -492,6 +492,32 @@ def DoMergeWrites(c1, c2):
     return ir, fwd
 
 
+def DoFoldIntoReduce(assign):
+    def access_to_str(node):
+        idx = f"[{','.join([str(idx) for idx in node.idx])}]" if node.idx else ""
+        return f"{node.name}{idx}"
+
+    assign_s = assign._node
+
+    if not isinstance(assign_s.rhs, LoopIR.BinOp) or assign_s.rhs.op != "+":
+        raise SchedulingError("The rhs of the assignment must be an add.")
+    if not isinstance(assign_s.rhs.rhs, LoopIR.Read) or access_to_str(
+        assign_s
+    ) != access_to_str(assign_s.rhs.rhs):
```

Comment:
RHS is being hardcoded here, this will not be able to fold `a = a + 1.0`, right? I understand that users can commute to have `a` on the RHS that's confusing IMO

Reply:
I originally wrote it to handle all sorts of cases. But then I realized there is a billion cases to handle. So, I decided to make it as concise and deterministic as possible in accordance with the primitives design.

We can build automations that can handle this arbitrary cases in the standard library: handle arbitrary reassociation and commute to put the expression in the right shape and then fold it.

Reply:
How about communicating that to users via primitive names? `reassociate_expr` and `fold_into_reduce` look like they are general primitives that work on arbitrary expressions. Maybe rename them `reassociate_rhs_expr` and `fold_rhs_into_reduce` or something like that?

Reply:
That's fair I think. I have been thinking about the same thing actually.

Reply:
I think the operation would be called `left_reassociate_expr` since this format is (a + b) + c means that addition has left associativity.

Reply:
I am not actually sure about this name `fold_rhs_into_reduce`. Because we pass a cursor to an assign statement. So, is the rhs for the assignment or the rhs of the expression that is on the rhs of the assignment.

Reply:
One option is to simply support both the LHS and the RHS of this operation.

Reply:
I think this name `fold_into_reduce` might be fine. `a = a + 1` -> `a += 1` wouldn't really be folding. While `a = 1 + a` -> `a += 1` would be folding. At least if you think about the physical movement of folding.

Reply:
It should be either "right most" or "left most" expressions, right?
Because the expression can be sth like `a = (3.0 + (2.0 + a))` in which case you'll want to traverse down to the right most expression

Reply:
Hmmm I might have been wrong about the canonical way in what direction to go:
https://www.ibm.com/docs/en/i/7.2?topic=operators-compound-assignment

Reply:
In the case you wrote about, you would reassociate first, then fold.

Reply:
I think I should change the rewrite to be `a = a + ...` -> `a += ...`. There is still a question of what to call it.

Reply:
There must be a canonical name for this rewrite but I can't find it anywhere.